### PR TITLE
Debugger - Eval_at

### PR DESCRIPTION
**Pull request by kisbg**
Signed-off-by: bence gabor kis <kisbg@inf.u-szeged.hu>

### CODE REVIEW

**Comment by zherczeg** [comment]
This feature will be needed to change variables in vscode.

File: src/debugger/Debugger.cpp [hunk 1]
```diff
@@ -166,20 +166,22 @@ bool Debugger::doEval(ExecutionState* state, ByteCodeBlock* byteCodeBlock, uint8
     uint8_t type = (uint8_t)(buffer[0] + 1);
     uint32_t size;
     char* data;
+    uint32_t index;
 
     memcpy(&size, buffer + 1, sizeof(uint32_t));
+    memcpy(&index, buffer + 1 + sizeof(uint32_t), sizeof(uint32_t));
```
Comment (zherczeg) [inline src/debugger/Debugger.cpp:172]: I think there is a minimum message size check somewhere, that should be updated as well.

File: src/debugger/Debugger.cpp [hunk 2]
```diff
@@ -166,20 +166,22 @@ bool Debugger::doEval(ExecutionState* state, ByteCodeBlock* byteCodeBlock, uint8
...
-        data = (char*)GC_MALLOC_ATOMIC(size);
+        data = (char*)GC_MALLOC_ATOMIC(size - sizeof(uint32_t));
         char* ptr = data;
-        char* end = data + size;
+        char* end = data + size - sizeof(uint32_t);
 
-        length = ESCARGOT_DEBUGGER_MAX_MESSAGE_LENGTH - 1 - sizeof(uint32_t);
-        memcpy(ptr, buffer + 1 + sizeof(uint32_t), length);
+        length = ESCARGOT_DEBUGGER_MAX_MESSAGE_LENGTH - 1 - sizeof(uint32_t) - sizeof(uint32_t);
+        memcpy(ptr, buffer + 1 + sizeof(uint32_t) + sizeof(uint32_t), length);
```
Comment (zherczeg) [inline src/debugger/Debugger.cpp:184]: To simplify source sending, maybe we could read the whole data, and get the index later (similar to extracting the file name in source sending). But if this is not a problem, just disregard this comment.

File: tools/debugger/debugger.py [hunk 3]
```diff
@@ -165,6 +165,29 @@ def do_eval(self, args):
...
+        try:
+            args = args.split(" ", 1)
+
+            index = int(args[0])
+
+            if len(args) == 2:
+                code = args[1]
+
+            if index < 0 or index > 65535:
+                raise ValueError("Invalid scope chain index: %d (must be between 0 and 65535)" % index)
```
Comment (zherczeg) [inline tools/debugger/debugger.py:182]: I think the limit can be UINT32_MAX, although we might never encounter that.

File: tools/debugger/debugger_core.py [hunk 4]
```diff
@@ -896,14 +900,17 @@ def _send_string(self, args, message_type):
...
+            args = struct.pack(self.byte_order + "I", index) + args
+
         size = len(args)
         max_fragment = min(self.max_message_size - message_header, size)
 
         message = struct.pack(self.byte_order + "BBI",
                               max_fragment + message_header,
                               message_type,
                               size)
-
```
Comment (zherczeg) [inline tools/debugger/debugger_core.py:906]: Don't remove this line.

**Comment by clover2123** [comment]
@kisbg There was no update in about a month. Is this pr still in WIP?

**Comment by kisbg** [comment]
@clover2123  The reason why is this pull request is not been updated, because there was a few other pull request (namely #671) that are needed for this feature.

File: src/debugger/Debugger.cpp [hunk 5]
```diff
@@ -262,10 +270,27 @@ bool Debugger::doEval(ExecutionState* state, ByteCodeBlock* byteCodeBlock, uint8
     type = ESCARGOT_MESSAGE_EVAL_RESULT_8BIT;
     m_stopState = ESCARGOT_DEBUGGER_IN_EVAL_MODE;
 
+
```
Comment (zherczeg) [inline src/debugger/Debugger.cpp:273]: No need this newline.

File: tools/debugger/debugger_core.py [hunk 6]
```diff
@@ -96,7 +96,6 @@
 ESCARGOT_DEBUGGER_PENDING_CONFIG = 19
 ESCARGOT_DEBUGGER_PENDING_RESUME = 20
 
-
```
Comment (zherczeg) [inline tools/debugger/debugger_core.py:99]: Keep this newline. It is python code.

File: tools/debugger/debugger_core.py [hunk 7]
```diff
@@ -703,6 +702,12 @@ def eval(self, code):
         self._send_string(code, ESCARGOT_MESSAGE_EVAL_8BIT_START)
         self.prompt = False
 
+    def eval_at(self, code, index):
+        self._send_string(code, ESCARGOT_MESSAGE_EVAL_8BIT_START, index)
+        self.prompt = False
+
+
+
```
Comment (zherczeg) [inline tools/debugger/debugger_core.py:710]: Only one newline.

File: src/debugger/Debugger.cpp [hunk 8]
```diff
@@ -262,10 +270,27 @@ bool Debugger::doEval(ExecutionState* state, ByteCodeBlock* byteCodeBlock, uint8
...
     try {
         Value asValue(str);
-        Value result(state->context()->globalObject()->evalLocal(*state, asValue, Value(Value::Undefined), byteCodeBlock->m_codeBlock->asInterpretedCodeBlock(), true));
-        str = result.toStringWithoutException(*state);
+        if (index != 0) {
```
Comment (zherczeg) [inline src/debugger/Debugger.cpp:276]: I think we only have lex env index, but we also need a state depth as well (see other functions for the correct name). First we go up in states, then in lex envs.

**Pull request closed without merging**